### PR TITLE
[type:feat]Add dynamically scale admin module

### PR DESCRIPTION
**Pull request by Terminator111**
Make sure that:

- [x] You have read the [contribution guidelines](https://shenyu.apache.org/community/contributor-guide).
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] Your local test passed `./mvnw clean install -Dmaven.javadoc.skip=true`.

### CODE REVIEW

**Comment by moremind** [comment]
not delete icon

**Comment by Terminator111** [comment]
> not delete icon

done.

**Review by misaya295 — COMMENTED** [review]
Empty tests may not be submitted

**Review by misaya295 — APPROVED** [review]

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+/**
+ * Test cases for ScalePolicyMapper.
+ */
+public final class ScalePolicyMapperTest extends AbstractSpringIntegrationTest {
+
+    @Resource
+    private ScalePolicyMapper scalePolicyMapper;
+
+    @Test
+    void deleteByPrimaryKey() {
```
Comment (misaya295) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java:36]: Please complete the unit test

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        // assertNotNull(resultDO);
+        assertEquals("1", resultDO.getId());
+        assertEquals(1, resultDO.getSort());
+        assertEquals(0, resultDO.getStatus());
+        assertEquals(10, resultDO.getNum());
+    }
+
+    @Test
+    void updateByPrimaryKeySelective() {
+    }
```
Comment (misaya295) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java:59]: same question

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+    @Test
+    void updateByPrimaryKey() {
+    }
```
Comment (misaya295) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java:63]: same question

File: shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    void deleteByPrimaryKey() {
+    }
+
+    @Test
+    void insert() {
+    }
+
+    @Test
+    void insertSelective() {
+    }
```
Comment (misaya295) [inline shenyu-admin/src/test/java/org/apache/shenyu/admin/mapper/ScalePolicyMapperTest.java:45]: same question

File: shenyu-admin/pom.xml [hunk 5]
```diff
@@ -352,6 +352,12 @@
             <artifactId>shenyu-admin-listener-zookeeper</artifactId>
             <version>${project.version}</version>
         </dependency>
+        <dependency>
+            <groupId>joda-time</groupId>
+            <artifactId>joda-time</artifactId>
```
Comment (misaya295) [inline shenyu-admin/pom.xml:357]: remove useless dependencies

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScalePolicyController.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    /**
+     * update scale policy.
+     *
+     * @param id primary key
+     * @param scalePolicyDTO scale policy info
+     * @return ShenyuAdminResult
+     */
+    @PutMapping("/{id}")
+    public ShenyuAdminResult updateScalePolicy(@PathVariable("id") @Valid
+                                               @Existed(provider = ScalePolicyMapper.class,
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScalePolicyController.java:83]: `@valid` should put on another line

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScalePolicyController.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                                                        message = "scale policy is not existed") final String id,
+                                               @Valid @RequestBody final ScalePolicyDTO scalePolicyDTO) {
+        scalePolicyDTO.setId(id);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScalePolicyController.java:86]: the id should in scalePolicyDTO

Comment (moremind) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScalePolicyController.java:86]: you can check in pathvaraiable and body

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    /**
+     * detail scale rule.
+     *
+     * @param id id
+     * @return ShenyuAdminResult
+     */
+    @GetMapping("/{id}")
+    public ShenyuAdminResult detailRule(@PathVariable("id") @Valid
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java:92]: `@valid` same question

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+     * @param id           id
+     * @param scaleRuleDTO scaleRuleDTO
+     * @return ShenyuAdminResult
+     */
+    @PutMapping("/{id}")
+    public ShenyuAdminResult updateRule(@PathVariable("id") @Valid
+                                        @Existed(provider = ScaleRuleMapper.class,
+                                                message = "scale rule is not existed") final String id,
+                                        @Valid @RequestBody final ScaleRuleDTO scaleRuleDTO) {
+        scaleRuleDTO.setId(id);
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java:124]: same question

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ScalePolicyDTO.java [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, sort, status, num, beginTime, endTime);
+    }
+
+    public static final class ScalePolicyDTOBuilder {
+
+        private @Existed(provider = ScalePolicyMapper.class, nullOfIgnore = true, message = "scale policy is not existed") String id;
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ScalePolicyDTO.java:226]: the annotation location is ugly

Comment (moremind) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ScalePolicyDTO.java:226]: `
@Existed(provider = ScalePolicyMapper.class, nullOfIgnore = true, message = "scale policy is not existed")
private String id;
`

**Review by moremind — CHANGES_REQUESTED** [review]

**Review by misaya295 — APPROVED** [review]

**Review by misaya295 — APPROVED** [review]

**Review by misaya295 — DISMISSED** [review]

**Comment by moremind** [comment]
you should add sql in upgrade script, for example, 2.6.1-2.7.0-mysql.sql, 2.6.1-2.7.0-pg.sql, and so on...

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+    /**
+     * query scale rule.
+     *
+     * @param metricName  metricName
+     * @param currentPage currentPage
+     * @param pageSize    pageSize
+     * @return org.apache.shenyu.admin.model.result.ShenyuAdminResult
+     */
+    @GetMapping("")
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java:76]: remove `("")`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+                .orElse(ShenyuAdminResult.error(ShenyuResultMessage.DETAIL_FAILED));
+    }
+
+    /**
+     * create scale rule.
+     *
+     * @param scaleRuleDTO scaleRuleDTO
+     * @return ShenyuAdminResult
+     */
+    @PostMapping("")
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java:108]: remove `("")`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+        return ShenyuAdminResult.success(ShenyuResultMessage.CREATE_SUCCESS, scaleRuleService.createOrUpdate(scaleRuleDTO));
+    }
+
+    /**
+     * update rule.
+     *
+     * @param scaleRuleDTO scaleRuleDTO
+     * @return ShenyuAdminResult
+     */
+    @PutMapping("")
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/controller/ScaleRlueController.java:119]: remove `("")`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ScalePolicyDTO.java [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    /**
+     * end time.
+     */
+    private String endTime;
+
+    public ScalePolicyDTO() {
+    }
+
+    public ScalePolicyDTO(final String id,
+                          @NotNull final Integer sort,
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/dto/ScalePolicyDTO.java:72]: remove `@Notnull`

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/entity/ScalePolicyDO.java [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+    public static ScalePolicyDO buildScalePolicyDO(final ScalePolicyDTO scalePolicyDTO) {
+        return Optional.ofNullable(scalePolicyDTO).map(item -> {
+            Timestamp currentTime = new Timestamp(System.currentTimeMillis());
+            try {
+                ScalePolicyDO scalePolicyDO =
+                        ScalePolicyDO.builder()
+                                .sort(item.getSort())
+                                .status(item.getStatus())
+                                .num(item.getNum())
+                                .beginTime(DateUtils.isValidDate(item.getBeginTime())
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/entity/ScalePolicyDO.java:239]: Why is the begintime null?

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ScalePolicyVO.java [hunk 16]
```diff
@@ -0,0 +1,298 @@
...
+    /**
+     * buildScalePolicyVO.
+     *
+     * @param scalePolicyDO scalePolicyDO
+     * @return ScalePolicyVO
+     */
+    public static ScalePolicyVO buildScalePolicyVO(final ScalePolicyDO scalePolicyDO) {
+        return Optional.ofNullable(scalePolicyDO)
+                .map(item -> {
+                    String beginTime = item.getBeginTime() != null
```
Comment (moremind) [inline shenyu-admin/src/main/java/org/apache/shenyu/admin/model/vo/ScalePolicyVO.java:249]: same question

File: shenyu-admin/src/main/java/org/apache/shenyu/admin/model/entity/ScalePolicyDO.java (hunk 15 above)
Comment (Terminator111) [reply shenyu-admin/src/main/java/org/apache/shenyu/admin/model/entity/ScalePolicyDO.java:239]: when use manual strategy or dynamic strategy, there is not need to fill it.

**Comment by Terminator111** [comment]
> you should add sql in upgrade script, for example, 2.6.1-2.7.0-mysql.sql, 2.6.1-2.7.0-pg.sql, and so on...

done.

**Review by misaya295 — CHANGES_REQUESTED** [review]

**Review by moremind — APPROVED** [review]

**Review by misaya295 — APPROVED** [review]

**Comment by misaya295** [comment]
@Terminator111 check ci pls

**Review by misaya295 — DISMISSED** [review]

**Review by misaya295 — APPROVED** [review]

**Pull request merged**